### PR TITLE
Overhaul 2.1 Remove Dependencies / Add Full Timm Support

### PR DESCRIPTION
**Pull request by isaaccorley**
This PR seeks to do a few things:

- remove the dependency on `pretrainedmodels` and `efficientnet-pytorch` as they are no longer maintained.
- remove the `mock` and `torchvision` dependency (torchvision isn't used anywhere anyway)
- Replace any lost model support by fully using timm (`torchseg.encoders.TimmEncoder`)
- add support for timm ViT encoders (torchseg.encoders.TimmViTEncoder`)
- Test all supported and unsupported timm encoders (they are defined here `torchseg.encoders.supported`). This includes ConvNext and Swin pretrained backbones
- Update tests to be more thorough
- Remove unecessary forward/backward calls in tests (this speeds up the tests significantly)

There's some other misc cleanup that is done:
- remove `Activation` class. We instead let the user choose the head activation when creating a model. Still defaults to `nn.Identity()`
- Allow for passing additional `encoder_params` to `timm.create_model(**kwargs)` in case users want to further customize the backbone

### CODE REVIEW

**Comment by JulienMaille** [comment]
Looking forward to test this update!
Have you checked if you can use timm `convenext_nano` encoder with this PR?

**Comment by isaaccorley** [comment]
> Looking forward to test this update!
> 
> Have you checked if you can use timm `convenext_nano` encoder with this PR?

Yes this PR allows the use of convnext as the encoder!

**Comment by isaaccorley** [comment]
@notprime Can you give this a review and maybe give some thoughts?

**Comment by ogencoglu** [comment]
Looking forward to this update...

File: torchseg/encoders/timm.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+            """
+            raise ValueError(err)
+
+        params = dict(
+            in_chans=in_channels,
+            features_only=True,
+            pretrained=pretrained,
+            out_indices=indices,
+            output_stride=output_stride,
+        )
```
Comment (notprime) [inline torchseg/encoders/timm.py:44]: Maybe we can add the custom `norm_layer` and `act_layer` directly here in `params` (and in `__init__`) if specified instead of leaving them in kwargs, just to make it clearer, what do you think? I just don't know how pretrained weights are handled in case of different norm and activation layers from the usual `BatchNorm2d` and `ReLU`, maybe we could force `pretrained` to False for now

**Review by notprime — CHANGES_REQUESTED** [review]
Everything seems fine to me, maybe we can specify act_layer and norm_layer directly in params instead of adding them through params.update via kwargs, just to make it clearer

File: torchseg/encoders/timm.py (hunk 1 above)
Comment (notprime) [inline torchseg/encoders/timm.py:44]: Maybe we can add the custom `norm_layer` and `act_layer` directly here in `params` (and in `__init__`) if specified instead of leaving them in kwargs, just to make it clearer, what do you think? I just don't know how pretrained weights are handled in case of different norm and activation layers from the usual `BatchNorm2d` and `ReLU`, maybe we could force `pretrained` to False for now

**Review by notprime — CHANGES_REQUESTED** [review]

File: torchseg/encoders/timm.py (hunk 1 above)
Comment (isaaccorley) [reply torchseg/encoders/timm.py:44]: Agreed I think this is a good addition. I think the only thing I need to check is just to make sure that act_layer and norm_layer are valid for most timm models. Can also pass these to the decoder as well.

Comment (notprime) [reply torchseg/encoders/timm.py:44]: Yes, forgot about the decoder.

I think most of timm models accept those parameters (except for old hard-coded models, like xception for instance, and specific transformer architectures which support only norm_layer), it could be a good idea to create a `supported_list` for models accepting both `act_layer` and `norm_layer` to avoid errors, or maybe directly add those to [torchseg/encoders/supported.py](https://github.com/isaaccorley/torchseg/pull/3/files#diff-63d2605567afb123ce55163d6d8f3e7b952d0fbcd9a1cda6b4e94131594c8c01) or to the json file created in [scripts/list_compatible_encoders.py](https://github.com/isaaccorley/torchseg/pull/3/files#diff-a6274f2a450efd1913eaf5f50a6bca64f6ac99dd30cb06351b458d3c9150cd7d)

Comment (isaaccorley) [reply torchseg/encoders/timm.py:44]: I think I'm going to punt on this for a future PR. Don't want this one to add too many new features than it already has.

**Review by notprime — APPROVED** [review]

**Comment by isaaccorley** [comment]
@JulienMaille @ogencoglu @notprime These features are now merged and you can install them using

`pip install --pre torchseg` or `pip install 'torchseg==0.0.1a2'`
